### PR TITLE
Search for methods with PEP 7 

### PR DESCRIPTION
**Pull request by Sondar74**
I added the PEP conditions in the `nonlinear_constraints.m` file. Now it is possible to search for methods with PEP up to 7 by adjusting the order `q` in this file. In my computer it takes a long time to run the code, so I asked Hendrick for help. The command is in the file `PEP_search.m` (this file can be deleted later). We want to make around 1000 searches with each algorithm. I am trying to find PEP(6,2,7) or PEP(7,2,7). 

Also, I want to ask how to keep track of the number of searches in this loop, since it looks like it may run forever unless it finds a solution. I made another loop to make the searches in parallel which allows me to keep track of the amount of searches, but I want to know if there is another optimal way.

### CODE REVIEW

**Comment by ketch** [comment]
> I want to ask how to keep track of the number of searches in this loop, since it looks like it may run forever unless it finds a solution. I made another loop to make the searches in parallel which allows me to keep track of the amount of searches, but I want to know if there is another optimal way.

The code automatically runs a number of searches equal to the number of parallel threads (`np`) multiplied by the number of starting points.  That's it.

How is the code for the q=7 conditions obtained?  Is it automated?  Can you share that code?

**Comment by Sondar74** [comment]
> > I want to ask how to keep track of the number of searches in this loop, since it looks like it may run forever unless it finds a solution. I made another loop to make the searches in parallel which allows me to keep track of the amount of searches, but I want to know if there is another optimal way.
> 
> The code automatically runs a number of searches equal to the number of parallel threads (`np`) multiplied by the number of starting points. That's it.
> 
> How is the code for the q=7 conditions obtained? Is it automated? Can you share that code?

I used two languages to generate the code: I get the conditions from `BSeries.jl` and I copy-paste them in a Python script to convert `julia -> Matlab`. I tested the current constraints with an RK7, and it satisfied all the constraints for `q = 7.` I will share the Jupyter Notebooks later, I am currently attending a meeting abroad.

**Comment by Sondar74** [comment]
@ketch @ranocha I added the code to generate the constraints in the past commit `Generate q=7 constraints`. Could you please take a look at them?

File: RK-coeff-opt/PEP_Generator/README.txt
Comment (ranocha) [inline RK-coeff-opt/PEP_Generator/README.txt:1]: Could you maybe also add the Julia `Project.toml` and `Manifest.toml` files that contain the versions of the packages that you used to generate the conditions?

File: RK-coeff-opt/PEP_Generator/README.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+1. The list of PEP conditions is obtained from 'Energy-preserving_conjugates.ipynb'
+
+2. This conditions must be applied to the file 'PEPGenerator.jl', which generates PEP conditions,
+one by one, in Butcher Notation.
+
+3. Finally, copy-paste the output in 'Translator.ipynb', which returns the PEP condition in Matlab format.
```
Comment (ranocha) [inline RK-coeff-opt/PEP_Generator/README.txt:6]:
```suggestion
3. Finally, copy-paste the output in 'Translator.ipynb', which returns the PEP condition in Matlab format.

This code was written for Julia v1.6. The versions of the Julia packages are recorded in the `Project.toml` and `Manifest.toml` files.
```

That's the version you used, isn't it?

Comment (Sondar74) [reply RK-coeff-opt/PEP_Generator/README.txt]: Yes, that's right.

**Comment by ketch** [comment]
@Sondar74 I'm finally going through the details of this PR; sorry it took me so long.  Here are my thoughts:

1.  I don't think this code should be merged into RK-Opt.  It should be kept in our energy preservation repository.
2.  Please don't make copies of files that are under version control; one of the main points of version control is to avoid having multiple copies of the same code that need to be maintained (I'm referring to `energy_preserving_conjugates.ipynb`).
3.  I don't understand your instructions.  You say "This conditions must be applied to the file 'PEPGenerator.jl', which generates PEP conditions, one by one, in Butcher Notation."  But I don't know how to "apply conditions to a file"; it seems like PEPGenerator.jl is a script with no inputs.  

Since it's all Julia code, please write one piece of code (it can be a file or a notebook) that generates all the conditions.

**Comment by Sondar74** [comment]
> @Sondar74 I'm finally going through the details of this PR; sorry it took me so long. Here are my thoughts:
> 
> 1. I don't think this code should be merged into RK-Opt.  It should be kept in our energy preservation repository.
> 2. Please don't make copies of files that are under version control; one of the main points of version control is to avoid having multiple copies of the same code that need to be maintained (I'm referring to `energy_preserving_conjugates.ipynb`).
> 3. I don't understand your instructions.  You say "This conditions must be applied to the file 'PEPGenerator.jl', which generates PEP conditions, one by one, in Butcher Notation."  But I don't know how to "apply conditions to a file"; it seems like PEPGenerator.jl is a script with no inputs.
> 
> Since it's all Julia code, please write one piece of code (it can be a file or a notebook) that generates all the conditions. We want to get rid of any manual steps, since those are likely to be sources of error.

Okey, for points 1 and 2, I will move the important files to the `energy preservation` repository. In particular, I will add the `nonlinear_constraints.m` and `Translator.ipynb`. From this, I got two questions:
1. Should I eliminate this PR?
2. Should I keep the `.toml` files?

As for the last point, I just added a new Julia Notebook in the energy preservarion repository where I have improved the Julia code to generate the conditions. Please take a look at it.

**Comment by ketch** [comment]
Yes, I think you should close this PR and put everything it contains, except for `nonlinear_contraints.m` into a separate PR.  Then just add the `nonlinear_constraints.m` changes to your branch here.  We'll figure out what to do with those later (I want to incorporate the energy-preserving conditions in a way that the user can choose whether to enforce them or not, and to what order).

**Pull request closed without merging**